aws/resource_aws_acm_certificate.go

```diff
@@ -77,10 +85,18 @@ func resourceAwsAcmCertificate() *schema.Resource {
 }
 
 func resourceAwsAcmCertificateCreate(d *schema.ResourceData, meta interface{}) error {
+	rand.Seed(time.Now().UnixNano())
 	acmconn := meta.(*AWSClient).acmconn
 	params := &acm.RequestCertificateInput{
 		DomainName:       aws.String(d.Get("domain_name").(string)),
-		ValidationMethod: aws.String(d.Get("validation_method").(string)),
+		IdempotencyToken: aws.String(strconv.Itoa(rand.Int())),
```

Comment:
@micbase Good question.

I was wondering if the goal of the token is to distinguish two (or possibly more) of the same calls to the `RequestCertificate` API (i.e. with the exact same parameters) so that the same certificate ARN is returned to prevent creating a new certificate if the client was somehow not able to read the response (e.g. network issues).
That is whether changing one parameter (e.g. adding a SAN) is considered to a different "request" from AWS perspective even if you provide the same idempotency token.

So, I run the following command against two PCAs in a single region from a single account during the same hour and got some **interesting** results.

Requesting a certificate multiple times with the same token:
```
$ date && aws acm request-certificate --domain-name pr-6666.com --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/1 --idempotency-token PR6666
Tue  5 Feb 2019 12:32:55 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/78d5b732-88a8-49af-8929-f28359eb8b75"
}
$ date && aws acm request-certificate --domain-name pr-6666.com --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/1 --idempotency-token PR6666
Tue  5 Feb 2019 12:33:01 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/78d5b732-88a8-49af-8929-f28359eb8b75"
}
```

Then adding a SAN yet keeping the same token will generate a new certificate:
```
$ date && aws acm request-certificate --domain-name pr-6666.com --subject-alternative-names pr-6666.org --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/1 --idempotency-token PR6666
Tue  5 Feb 2019 12:35:16 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/b75395d9-d9fe-4249-8b30-c7ac4d1fbfd0"
}

$ date && aws acm request-certificate --domain-name pr-6666.com --subject-alternative-names pr-6666.org --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/1 --idempotency-token PR6666
Tue  5 Feb 2019 12:35:32 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/b75395d9-d9fe-4249-8b30-c7ac4d1fbfd0"
}
```

Then changing the certificate authority (same account, same region) somehow returns the same certificate (**I consider that behaviour a bug**):
```
$ date && aws acm request-certificate --domain-name pr-6666.com --subject-alternative-names pr-6666.org --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/2 --idempotency-token PR6666
Tue  5 Feb 2019 12:36:05 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/b75395d9-d9fe-4249-8b30-c7ac4d1fbfd0"
}
```

Then adding one more SAN generates a new certificate:
```
$ date && aws acm request-certificate --domain-name pr-6666.com --subject-alternative-names pr-6666.org pr-6666.net --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/2 --idempotency-token PR6666
Tue  5 Feb 2019 12:39:03 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/0d41f00a-25ce-492e-b6c4-1bb34b64da60"
}
$ date && aws acm request-certificate --domain-name pr-6666.com --subject-alternative-names pr-6666.org pr-6666.net --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/2 --idempotency-token PR6666
Tue  5 Feb 2019 12:39:10 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/0d41f00a-25ce-492e-b6c4-1bb34b64da60"
}
```

And switching back to the original certificate authority do keep the same certificate (**again kind of unexpected behaviour IMHO**):
```
$ date && aws acm request-certificate --domain-name pr-6666.com --subject-alternative-names pr-6666.org pr-6666.net --certificate-authority-arn arn:aws:acm-pca:region:1234:certificate-authority/1 --idempotency-token PR6666
Tue  5 Feb 2019 12:39:30 PST
{
    "CertificateArn": "arn:aws:acm:region:1234:certificate/0d41f00a-25ce-492e-b6c4-1bb34b64da60"
}
```

I would be keen to only have the domain name as part of the hash but it seems that adding the PCA ARN could work around that bug of getting the same certificate.
However this might be fixed eventually (if it is a bug and not a feature).

What do you think?

Reply:
That is the exact problem I encountered, apologize for not expressing it clearly in my initial PR.
IMO, using two different PCAs should generate different certs, so I used IdempotencyToken to work it around. I'm not sure if that's a bug or feature, more like a bug to me.